keps/sig-auth/2784-csr-duration/README.md

```diff
@@ -0,0 +1,518 @@
+  //   3. Signer whose configured minimum is longer than the requested duration
+  //
+  // The minimum valid value for expirationSeconds is 600, i.e. 10 minutes.
+  //
+  // +optional
+  ExpirationSeconds *int32 `json:"expirationSeconds,omitempty" protobuf:"varint,8,opt,name=expirationSeconds"`
+
+  // go doc omitted for brevity
+  Usages []KeyUsage `json:"usages,omitempty" protobuf:"bytes,5,opt,name=usages"`
+
+  // ... other fields omitted for brevity
+}
+```
+
+The name `expirationSeconds` was chosen to match existing art in the token request
+API.  Similarly, the minimum valid duration was chosen to match the token request
+API as well.  As this is a security related field, individuals may be encouraged
+to set this value to the minimum valid value to maximize security.  Since a certificate
+with a short lifetime will require frequent rotation before the current certificate
+expires (say at `80%` the lifetime of the certificate given that CSR approval
+and signing is asynchronous which necessitates a buffer), `10` minutes seems like an
+appropriate minimum to prevent accidental DOS against the CSR API.  Furthermore,
+`10` minutes is a short enough lifetime that revocation is not of concern.
+
+Metrics will be included to show if requested expirations are being extended or
+truncated (i.e. is the requested duration being honored by the signer).
+
+### Test Plan
+
+Unit tests covering:
+
+1. Validation logic for minimum duration
+2. `pkg/controller/certificates/authority.PermissiveSigningPolicy` updates to handle `expirationSeconds`
+3. Metrics
+4. CSR REST storage ignores `spec.expirationSeconds` when the feature gate is disabled
+
+Integration test covering:
+
+1. Creating and approving CSRs and asserting that certificate signer controllers such as
+   `pkg/controller/certificates/signer.NewKubeAPIServerClientCSRSigningController` honor
+   `spec.expirationSeconds` by checking the duration of the issued certificate.
+
+### Graduation Criteria
+
+#### Alpha
+
+- This design will start at the beta phase and the functionality will be enabled by default
+
+This design represents a small, optional change to an existing GA API.  Thus it
+prioritizes rollout speed to allow clients to start using this functionality
+sooner (to reap potential security benefits) at the cost of data durability
+during version skews (discussed below).
+
+#### Beta
+
+- Feature fully implemented as described in this design
+- Unit tests completed and enabled
+- Integration test completed and enabled
+- [CSR docs](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests) updated with details about usage of the `spec.expirationSeconds` field
+
+#### GA
+
+- Confirm with [cert-manager](https://github.com/jetstack/cert-manager/pull/3646) that the new functionality addresses their use case
+- Confirm with [pinniped](https://pinniped.dev) that the new functionality addresses their use case
```

Comment:
non-blocking for merge of this doc, but can you confirm with them this addresses their use case ~now, not in a release?

Reply:
This looks good for the Pinniped use case as well. Our planned usage is for a credential exchange API just like the second user story:

1. External user authenticates to a trusted service (using some custom application-specific credential).
1. Trusted service creates a CSR with a short `spec.expirationSeconds` (probably the 10 minute minimum).
1. Trusted service approves its own CSR and relies on a default signer to sign it.
1. Trusted service retrieves the signed certificate and returns it to the authenticated external user.

We would quickly add support for new flow but keep some fallback code that uses a custom signer until we no longer support current Kubernetes versions. We can probe for this feature dynamically and select the modern or fallback implementation appropriately.